typescript/src/proof.ts

```diff
@@ -70,3 +79,253 @@ function createMerkleProof(txMerkleBranch: TransactionMerkleBranch): string {
+ *        to determine the path to traverse in the Merkle tree.
+ * @throws {Error} If the intermediate nodes are of an invalid length or if the
+ *         computed root hash does not match the merkle root hash parameter.
+ * @returns An empty return value.
+ */
+function validateMerkleTreeHashes(
+  transactionHash: string,
+  intermediateNodesHashes: string,
+  merkleRootHash: string,
+  transactionIndex: number
+) {
+  if (
+    intermediateNodesHashes.length === 0 ||
+    intermediateNodesHashes.length % 64 !== 0
+  ) {
+    throw new Error("Invalid merkle tree")
+  }
+
+  let idx = transactionIndex
+  let current = transactionHash
+
+  // i moves in increments of 64
+  for (let i = 0; i < intermediateNodesHashes.length; i += 64) {
+    if (idx % 2 === 1) {
+      current = computeHash256(
+        intermediateNodesHashes.slice(i, i + 64) + current
+      ).toString()
+    } else {
+      current = computeHash256(
+        current + intermediateNodesHashes.slice(i, i + 64)
+      ).toString()
+    }
+    idx = idx >> 1
+  }
+
+  if (current !== merkleRootHash) {
+    throw new Error(
+      "Transaction Merkle proof is not valid for provided header and transaction hash"
+    )
+  }
+}
+
+/**
+ * Validates a chain of consecutive block headers by checking each header's
+ * difficulty, hash, and continuity with the previous header. This function can
+ * be used to validate a series of Bitcoin block headers for their validity.
+ * @param blockHeaders An array of block headers that form the chain to be
+ *        validated.
+ * @param previousEpochDifficulty The difficulty of the previous Bitcoin epoch.
+ * @param currentEpochDifficulty The difficulty of the current Bitcoin epoch.
+ * @dev The block headers must come from Bitcoin epochs with difficulties marked
+ *      by the previous and current difficulties. If a Bitcoin difficulty relay
+ *      is used to provide these values and the relay is up-to-date, only the
+ *      recent block headers will pass validation. Block headers older than the
+ *      current and previous Bitcoin epochs will fail.
+ * @throws {Error} If any of the block headers are invalid, or if the block
+ *         header chain is not continuous.
+ * @returns An empty return value.
+ */
+function validateBlockHeadersChain(
+  blockHeaders: BlockHeader[],
+  previousEpochDifficulty: BigNumber,
+  currentEpochDifficulty: BigNumber
+) {
+  let requireCurrentDifficulty: boolean = false
+  let previousBlockHeaderHash: Hex = Hex.from("00")
+
+  for (let index = 0; index < blockHeaders.length; index++) {
+    const currentHeader = blockHeaders[index]
+
+    // Check if the current block header stores the hash of the previously
+    // processed block header. Skip the check for the first header.
+    if (index !== 0) {
+      if (
+        !previousBlockHeaderHash.equals(currentHeader.previousBlockHeaderHash)
+      ) {
+        throw new Error("Invalid headers chain")
+      }
+    }
```

Comment:
We are basically checking the chain integrity here, right?

Reply:
Yes.
The block headers need to be "connected": each block should point to the previously processed block header (except for the first one).